### PR TITLE
Support LeddarOne lidar

### PR DESCRIPTION
**Pull request by zehortigoza**
https://leddartech.com/modules/leddarone/

Here 2 videos showing the performance of this lidar in Aero:

- Holding position: https://www.youtube.com/watch?v=p9aTtwIk2-U

- Moving: https://www.youtube.com/watch?v=sMrkOJoF-jM

Flight log of the 2 videos: https://logs.px4.io/plot_app?log=78479503-532f-403a-b47b-c474054b746c

### CODE REVIEW

**Comment by hamishwillee** [comment]
@zehortigoza Can we add something for this in the docs? 

Specifically [Rangefinders](https://docs.px4.io/en/sensor/rangefinders.html) should have an entry explaining  what the rangefinder offers and how to enable it. 

IF this won't be in any configurations and requires specific extra work to get going it might be worth creating a separate page below rangefinders as well ([like this one for ulanding](https://docs.px4.io/en/sensor/ulanding_radar.html)). That would cover where/how you add the driver, how you make sure it is started, any additional configuration.

Thoughts?

**Comment by zehortigoza** [comment]
@hamishwillee Sure I will send a PR adding it to the documentation. I'm enabling it for AeroFC at least but I will also add instructions of how enable it for other boards.

**Comment by hamishwillee** [comment]
Thank you very much!

**Comment by zehortigoza** [comment]
@LorenzMeier You mean the tuning parameter in LeddarOne? Yes, I'm using the stock parameters.

**Comment by hamishwillee** [comment]
Thanks for the docs @zehortigoza ! I have added a note up the top that this isn't in master yet: https://docs.px4.io/en/sensor/leddar_one.html
Will remove and do a final tidy/check when this is merged.

File: src/drivers/leddar_one/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+# ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+#
+############################################################################
+px4_add_module(
+	MODULE drivers__leddar_one
+	MAIN leddar_one
+	STACK_MAIN 1200
+	COMPILE_FLAGS
+		-Os
```
Comment (dagar) [inline src/drivers/leddar_one/CMakeLists.txt]: Optimization level is already set for the build overall.

**Comment by dagar** [comment]
The parameter usage is a bit unusual (board specific and placing it within the board support). We'd probably be better off shoving the baudrate setting into the new driver itself.

Longer term I want to generalize the serial port setup. https://github.com/PX4/Firmware/issues/8302

File: src/drivers/leddar_one/CMakeLists.txt (hunk 1 above)
Comment (zehortigoza) [reply src/drivers/leddar_one/CMakeLists.txt]: Ok, I will remove it

**Comment by zehortigoza** [comment]
> The parameter usage is a bit unusual (board specific and placing it within the board support). We'd probably be better off shoving the baudrate setting into the new driver itself.
>
> Longer term I want to generalize the serial port setup. #8302

That would be nice but it is target for 1.8v right?! Not sure if the merge window for 1.7v is still open for this.
If it is still open we could merge this as is(if the rest is in good shape for you) and then move to #8302 .
AeroFC is having some problems with Garmin Lidar(and other I2C sensors) and we would like to have that merged to recommend costumers to use this one instead.

**Comment by mhkabir** [comment]
What is the reason for the large, increasing timeslip in the logs?

**Comment by zehortigoza** [comment]
> What is the reason for the large, increasing timeslip in the logs?
Huum not sure yet but a quick test showed that I have this timeslip with our without this patches.

**Comment by dagar** [comment]
My only issue with this is putting params in the board support, but I don't have a better suggestion at the moment.

**Comment by dagar** [comment]
Ok, how about this. All the other lidar units follow the same pattern of being enabled by a param and then started in rc.sensors. 
https://github.com/PX4/Firmware/blob/master/ROMFS/px4fmu_common/init.d/rc.sensors#L351
https://github.com/PX4/Firmware/blob/master/src/modules/sensors/sensor_params.c#L238

If you did this for the LeddarOne, then in the special aero mavlink handling you could either check the param, or even check if the LeddarOne is actually running.

You could then also add the leddar_one module across the board (px4fmu-v3, v4, v4pro, v5, auav-x21, mindpx).

**Comment by hamishwillee** [comment]
I upvote https://github.com/PX4/Firmware/pull/8409#issuecomment-350816653

From a user point of view it would be far more consistent to be able to say that all sensors are enabled by setting a parameter. From a "guidance to developers" point of view it would be great to have a single mandated way to make the integration work (provided that way doesn't badly limit what can be done)
Some thoughts:
- We have multiple serial ports but only one parameter. How does/would PX4 know that you have a rangefinder on a particular port? (I presume this does not matter for I2C which can have multiple things on bus)
- What happens if a ROM is built without the rangefinder driver but the parameter is set.
- Can/should we do something about misconfiguration - multiple parameters set, when only one thing can be on the port?

**Comment by dagar** [comment]
@hamishwillee very good points, we can try to address some of them in https://github.com/PX4/Firmware/issues/8302. For now I'm simply pushing for consistency.

**Comment by LorenzMeier** [comment]
Jenkins test this please

**Comment by zehortigoza** [comment]
Sorry for the delay I was in vacation.

> Ok, how about this. All the other lidar units follow the same pattern of being enabled by a param and then started in rc.sensors.
https://github.com/PX4/Firmware/blob/master/ROMFS/px4fmu_common/init.d/rc.sensors#L351
https://github.com/PX4/Firmware/blob/master/src/modules/sensors/sensor_params.c#L238
>
> If you did this for the LeddarOne, then in the special aero mavlink handling you could either check the param, or even check if the LeddarOne is actually running.
> 
> You could then also add the leddar_one module across the board (px4fmu-v3, v4, v4pro, v5, auav-x21, mindpx).

I agree with the idea to add a parameter to start leddar_one driver but the idea to check if a UART lidar parameter was set to not start MAVLink sounds like a bad hack that will break easily.
The ideal would https://github.com/PX4/Firmware/issues/8302 but until then for all other boards user can change SYS_COMPANION to disable MAVLink and for AeroFC user can change AEROFC_TELE_BPS. Maybe give a better name to this parameter like: AEROFC_EXT_TELE or AEROFC_TELEMETR

Let me know what you guys think so I can change/fix and rebase this.

**Comment by LorenzMeier** [comment]
@ChristophTobler Could you chime in here? @zehortigoza It's really hard for me to predict the tradeoffs without testing it - could you do your best guess and rebase the PR?

File: src/drivers/leddar_one/leddar_one.cpp [hunk 2]
```diff
@@ -0,0 +1,564 @@
...
+		if (report.current_distance > report.min_distance
+		    && report.current_distance < report.max_distance) {
+			valid = true;
+		}
+
+		printf("valid: %u\n", valid);
+		printf("distance: %0.3fm\n", (double)report.current_distance);
+		printf("time: %llu\n", report.timestamp);
+
+	} else {
```
Comment (ChristophTobler) [inline src/drivers/leddar_one/leddar_one.cpp]: There is no way of stopping the driver

**Comment by ChristophTobler** [comment]
I would also push for consistency and add a parameter to enable it

**Comment by ChristophTobler** [comment]
@zehortigoza Is the Aero in this video using the optical flow app that we did? It holds position nicely :+1:! Did you do any other modifications except dampening?

**Comment by zehortigoza** [comment]
Okay I will fix and rebase it until monday.

> @zehortigoza Is the Aero in this video using the optical flow app that we did? It holds position nicely 👍! Did you do any other modifications except dampening?

Yes it the aero-optical-flow.
Yep just the dampening but I guess for position hold it would work as good with the stock setup too.

**Comment by hamishwillee** [comment]
@zehortigoza Ping - any plan to fix and rebase?

**Comment by zehortigoza** [comment]
Rebased and fixed following the same approach did to Benewake TFMini.
Sorry for the delay

File: src/drivers/distance_sensor/leddar_one/leddar_one.cpp [hunk 3]
```diff
@@ -0,0 +1,564 @@
...
+		usleep(1000);
+	}
+
+	PX4_ERR("No readings from " NAME);
+
+	return PX4_ERROR;
+}
+
+void
+leddar_one::stop()
```
Comment (ChristophTobler) [inline src/drivers/distance_sensor/leddar_one/leddar_one.cpp]: I don't think this is being executed, right?

Comment (zehortigoza) [reply src/drivers/distance_sensor/leddar_one/leddar_one.cpp]: Fixed

**Comment by zehortigoza** [comment]
@ChristophTobler Fixed almost all CI errors, there is just one now caused by lack of flash memory in px4fmu-v2, should I remove the rangesensor's from px4fmu-v2?

**Comment by ChristophTobler** [comment]
@zehortigoza Instead of removing all range finders we could add them individually.

**Comment by ChristophTobler** [comment]
@zehortigoza I wouldn't remove all of them because this might brake some people's setup. Can you add them individually? `drivers/distance_sensor/ll40ls` etc.

**Comment by zehortigoza** [comment]
@ChristophTobler Okay fixed and now all CI tests are passing.

**Pull request merged**